plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java

```diff
@@ -0,0 +1,516 @@
+                return o2.getUpdatedOn().compareTo(o1.getUpdatedOn()); // desc
+            }
+        });
+
+        boolean have_balance_entries = false;
+        //check that there is at least one balance entry
+        for (Iterator<QuotaBalanceVO> it = quotaBalance.iterator(); it.hasNext();) {
+            QuotaBalanceVO entry = it.next();
+            if (entry.getCreditsId() > 0) {
+                have_balance_entries = true;
+                break;
+            }
+        }
+        //if last entry is a credit deposit then remove that as that is already
+        //accounted for in the starting balance after that entry, note the sort is desc
+        if (have_balance_entries) {
+            ListIterator<QuotaBalanceVO> li = quotaBalance.listIterator(quotaBalance.size());
+            // Iterate in reverse.
+            while (li.hasPrevious()) {
+                QuotaBalanceVO entry = li.previous();
+                if (s_logger.isDebugEnabled()) {
+                    s_logger.debug("createQuotaBalanceResponse: Entry=" + entry);
+                }
+                if (entry.getCreditsId() > 0) {
+                    li.remove();
+                } else {
+                    break;
+                }
+            }
+        }
+
+        int quota_activity = quotaBalance.size();
+        QuotaBalanceResponse resp = new QuotaBalanceResponse();
+        BigDecimal lastCredits = new BigDecimal(0);
+        boolean consecutive = true;
+        for (Iterator<QuotaBalanceVO> it = quotaBalance.iterator(); it.hasNext();) {
+            QuotaBalanceVO entry = it.next();
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug("createQuotaBalanceResponse: All Credit Entry=" + entry);
+            }
+            if (entry.getCreditsId() > 0) {
+                if (consecutive) {
+                    lastCredits = lastCredits.add(entry.getCreditBalance());
+                }
+                resp.addCredits(entry);
+                it.remove();
+            } else {
+                consecutive = false;
+            }
+        }
+
+        if (quota_activity > 0 && quotaBalance.size() > 0) {
+            // order is desc last item is the start item
+            QuotaBalanceVO startItem = quotaBalance.get(quotaBalance.size() - 1);
+            QuotaBalanceVO endItem = quotaBalance.get(0);
+            resp.setStartDate(startItem.getUpdatedOn());
+            resp.setStartQuota(startItem.getCreditBalance());
+            resp.setEndDate(endItem.getUpdatedOn());
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug("createQuotaBalanceResponse: Start Entry=" + startItem);
+                s_logger.debug("createQuotaBalanceResponse: End Entry=" + endItem);
+            }
+            resp.setEndQuota(endItem.getCreditBalance().add(lastCredits));
+        } else if (quota_activity > 0) {
+            // order is desc last item is the start item
+            resp.setStartDate(startDate);
```

Comment:
Since ``Date``s are mutable, ``startDate`` should be copied before being passed into ``setStartDate``.